### ISSUE TITLE
Results missmatch for model order selection and CCRB

### ISSUE THREAD

**Issue by SSPzhulingke** [ShlezingerLab/AI-Subspace-Methods#1]
To be exact, models trained with "threshold" perform best that they can reach over 90% performance while "mdl" and "aic" perform terribly.
Currently I don"t figure out the reasons making for the result, so I truly consult with you about any suggestions to re-implement your great project successfully.

**Comment by aradgast** [comment]
Hi @Illgotry,
Please provide more details about the simulation you ran and the training process, including the simulation and training parameters.

**Comment by SSPzhulingke** [comment]
> Hi [@Illgotry](https://github.com/Illgotry), Please provide more details about the simulation you ran and the training process, including the simulation and training parameters.

Thanks for your sincere reply!
I didn't modify your code generally and merely modified the simulation_commands and regularization type of SubspaceNet in the "main.py", part of code are below.
Then I ran the "main.py" with None, "aic", "mdl" and "threshold" respectively, obtaining weights file like"NFSubspaceNet_tau=8_diff_method=2d_music_field_type=near_train_loss_type=music_spectrum_threshold_N=15_M=2_T=100_narrowband_SNR=10_near_field_non-coherent_eta=0.0_sv_var=0.0.pt".
Then their evalution results are just like what I described before.

`# Imports
import os
import warnings
import time
import matplotlib.pyplot as plt
from run_simulation import run_simulation
import argparse

#Initialization
#os.system("cls||clear")
plt.close("all")

scenario_dict = {
    # "SNR": [-10, -5, 0, 5, 10],
    # "T": [10, 20, 30, 50, 70, 100],
    #  "eta": [0.0, 0.01, 0.02, 0.03, 0.04, 0.05],
    # "M": [2, 3, 4, 5, 6, 7],
}

simulation_commands = {
    "SAVE_TO_FILE": False,
    "CREATE_DATA": True,
    "SAVE_DATASET": False,
    "LOAD_MODEL": False,
    "TRAIN_MODEL": True,
    "SAVE_MODEL": True,
    "EVALUATE_MODE": True,
    "PLOT_RESULTS": True,  # if True, the learning curves will be plotted
    "PLOT_LOSS_RESULTS": True,  # if True, the RMSE results of evaluation will be plotted
    "PLOT_ACC_RESULTS": True,  # if True, the accuracy results of evaluation will be plotted
    "SAVE_PLOTS": False,  # if True, the plots will be saved to the results folder
}

system_model_params = {
    "N": 15,  # number of antennas
    "M": 2,  # number of sources
    "T": 100,  # number of snapshots
    "snr": 10,  # if defined, values in scenario_dict will be ignored
    "field_type": "near",  # Near, Far
    "signal_type": "Narrowband",  # Narrowband, broadband
    "signal_nature": "non-coherent",  # if defined, values in scenario_dict will be ignored
    "eta": 0.0,  # steering vector uniform error variance with respect to the wavelength.
    "bias": 0, # steering vector bias error
    "sv_noise_var": 0.0, # steering vector addative gaussian error noise variance
    "doa_range": 60, # The range of the DOA values [-doa_range, doa_range]
    "doa_resolution": .5, # The resolution of the DOA values in degrees
    "max_range_ratio_to_limit": 0.5, # The ratio of the maximum range in respect to the Fraunhofer distance
    "range_resolution": 1, # The resolution of the range values in meters
    "wavelength": 1, # The carrier wavelength of the signal in meters
}
model_config = {
    "model_type": "SubspaceNet",  # SubspaceNet, DCD-MUSIC, DeepCNN, TransMUSIC, DR_MUSIC
    "model_params": {}
}
if model_config.get("model_type") == "SubspaceNet":
    model_config["model_params"]["diff_method"] = "music_2D"  # esprit, music_1D, music_2D, beamformer
    model_config["model_params"]["train_loss_type"] = "music_spectrum"  # music_spectrum, rmspe, beamformerloss
    model_config["model_params"]["tau"] = 8
    model_config["model_params"]["field_type"] = "Near"  # Far, Near
    model_config["model_params"]["regularization"] = "threshold" # aic, mdl, threshold, None
    model_config["model_params"]["variant"] = "small"  # big, small
    model_config["model_params"]["norm_layer"] = True
    model_config["model_params"]["batch_norm"] = False

elif model_config.get("model_type") == "DCD-MUSIC":
    model_config["model_params"]["tau"] = 8
    model_config["model_params"]["diff_method"] = ("esprit", "music_1D")  # ("esprit", "music_1D")
    #model_config["model_params"]["train_loss_type"] = ("rmspe", "rmspe")  # ("rmspe", "rmspe"), ("rmspe",
    # "music_spectrum"), ("music_spectrum", "rmspe")
    model_config["model_params"]["regularization"] = None  # aic, mdl, threshold, None
    model_config["model_params"]["variant"] = "small"  # big, small
    model_config["model_params"]["norm_layer"] = True

elif model_config.get("model_type") == "DeepCNN":
    model_config["model_params"]["grid_size"] = 361

training_params = {
    "samples_size": 4096,
    "train_test_ratio": 0.1,
    "training_objective": "angle, range",  # angle, range, source_estimation
    "batch_size": 128,
    "epochs": 50,
    "optimizer": "Adam",  # Adam, SGD
    "scheduler": "ReduceLROnPlateau",  # StepLR, ReduceLROnPlateau
    "learning_rate": 0.001,
    "weight_decay": 1e-9,
    "step_size": 50,
    "gamma": 0.5,
    "true_doa_train": None,  # if set, this doa will be set to all samples in the train dataset
    "true_range_train": None,  # if set, this range will be set to all samples in the train dataset
    "true_doa_test": None,  # if set, this doa will be set to all samples in the test dataset
    "true_range_test": None,  # if set, this range will be set to all samples in the train dataset
    "use_wandb": False,
    "simulation_name": None,
}
evaluation_params = {
    "models": {
        # "TransMUSIC": {
        #                 "model_name": "TransMUSIC",
        #             },
        # "DCD-MUSIC": {
        #             "model_name": "DCD-MUSIC",
        #             "tau": 8,
        #             "diff_method": ("esprit", "music_1D"),
        #             #"train_loss_type":("rmspe", "rmspe"),
        #             "regularization": "mdl",
        #               },
        #  "DCD-MUSIC_V2": {
        #             "model_name": "DCD-MUSIC",
        #             "tau": 8,
        #             "diff_method": ("esprit", "music_1d"),
        #             "regularization": "aic",
        #             "variant": "big"
        #               },
        "NFSubspaceNet": {
                        "model_name": "SubspaceNet",
                        "tau": 8,
                        "diff_method": "music_2D",
                        "train_loss_type": "music_spectrum",
                        "field_type": "near",
                        "regularization": "threshold",
                         },
        # "NFSubspaceNet_V2": {
        #                 "model_name": "SubspaceNet",
        #                 "tau": 8,
        #                 "diff_method": "music_2D",
        #                 "train_loss_type": "music_spectrum",
        #                 "field_type": "near",
        #                 "regularization": "threshold",
        #                 "variant": "big",
        #                 },
    },
    "augmented_methods": [
        # ("SubspaceNet", "beamformer", {"tau": 8, "diff_method": "music_2D", "train_loss_type": "music_spectrum", "field_type": "near"}),
        # ("SubspaceNet", "beamformer", {"tau": 8, "diff_method": "esprit", "train_loss_type": "rmspe", "field_type": "far"}),
        # ("SubspaceNet", "esprit", {"tau": 8, "diff_method": "esprit", "train_loss_type": "rmspe", "field_type": "far"}),
    ],
    "subspace_methods": [
        # "CCRB",
         "2D-MUSIC",
         "Beamformer",
        # "CS_Estimator",
        # "ESPRIT",
        # "1D-MUSIC",
        # "Root-MUSIC",
        # "TOPS",
    ]
}`

**Comment by aradgast** [comment]
@Illgotry Thanks for the clarification.
I re-run the simulation locally, using AIC test for training and testing.
For the best stable results, i would suggest adding an extra norm layer before the gramm overloading tranformation(see [here](https://github.com/ShlezingerLab/AI-Subspace-Methods/blob/e061bed308cae9fb99051c474687163ac814869e/src/models_pack/subspacenet.py#L116)).
This is the results i got for 4K training samples(90% for training and 10% for validation) with batch size of 32 and for 50 epochs(on RTX 6000), test over 400 samples:
1. norm layer after gramm(from the original paper): 85.25%
2. norm layer before gramm: 96.7% (with thresholding got 93.5% but a bit better localization results)
3. norm layer before and after: 97%


let me know how it works for you.

Thanks for your feedback!

**Comment by SSPzhulingke** [comment]
> [@Illgotry](https://github.com/Illgotry) Thanks for the clarification. I re-run the simulation locally, using AIC test for training and testing. For the best stable results, i would suggest adding an extra norm layer before the gramm overloading tranformation(see [here](https://github.com/ShlezingerLab/AI-Subspace-Methods/blob/e061bed308cae9fb99051c474687163ac814869e/src/models_pack/subspacenet.py#L116)). This is the results i got for 4K training samples(90% for training and 10% for validation) with batch size of 32 and for 50 epochs(on RTX 6000), test over 400 samples:
> 
> 1. norm layer after gramm(from the original paper): 85.25%
> 2. norm layer before gramm: 96.7% (with thresholding got 93.5% but a bit better localization results)
> 3. norm layer before and after: 97%
> 
> let me know how it works for you.
> 
> Thanks for your feedback!

Wow! After adding a norm layer before gramm and modifying the batch size from 128 to 32, the accuracy of NFSubspaceNet using AIC reaches 97% indeed.I think perhaps it's some parameter that influences the performance.I'll try to re-train other models.
Anyway,thanks a lot for your guidance!!

**Comment by aradgast** [comment]
Great.
Fill free to reach out if you have any more questions.

**Comment by SSPzhulingke** [comment]
Sorry to bother you again.
the "crrb" values I got which should correspond to the Figure 4(a) in your article"Near Field Localization via AI-Aided Subspace Methods", are always around 0.XX. For instance, when SNR=-10dB, the RMSPE of CCRB should be 10 approximately while my value is still near 0.5.
I checked the code but cannot figure it out.
Another question is about Figure 4(b). DCDMUSIC using "aic" I trained in 5dB and 10dB performs bad. To be exact,the accuracy of source-estimation is 0 and RMSPE is higher than the one in 0dB.
When you are free, I'm looking forward to your instrutions!

**Comment by aradgast** [comment]
@Illgotry No worries about bothering me, if you find bugs in my code, that would be great!

For your first issue, please see the cited reference for the implementation of this bound value.
M. N. El Korso, R. Boyer, A. Renaux, and S. Marcos, “Conditional and
 unconditional Cram´er–Rao bounds for near-field source localization,”
 IEEE Trans. Signal Process., vol. 58, no. 5, pp. 2901–2907, 2010
If you find any bugs in the code, let me know. (You can also open a PR if you have a fix.)

For your second issue, I would suggest trying to adjust the regularization coefficient, which naturally varies across different tests and models. I don’t recall the exact configuration, but let me know if you continue to struggle with it.

**Comment by aradgast** [comment]
@Illgotry I have reviewed my code implementation and it's appears that i mistakenly squared the number of sensors in an incorrect location in [here](https://github.com/ShlezingerLab/AI-Subspace-Methods/blob/1114656744c38ea7f6ec08a7d2a5bcc365f0ce3d/src/evaluation.py#L421). 
After removing this error and reproducing the simulation setup from [72], the results seem to be aligned.
In any case, this bug improves the CCRB results but does not affect our proposed algorithm.
Please approve so i will merge this fix.

**Comment by SSPzhulingke** [comment]
I agree the CCRB implementation is correct after correction, but the CCRB results are still far away from Figure 4(a) in low SNRs somehow.
Besides, my second issue is solved after your advice. Appreciate !!

**Comment by aradgast** [comment]
> I agree the CCRB implementation is correct after correction, but the CCRB results are still far away from Figure 4(a) in low SNRs somehow. Besides, my second issue is solved after your advice. Appreciate !!

I just noticed the issue — see [here](https://github.com/ShlezingerLab/AI-Subspace-Methods/blob/1114656744c38ea7f6ec08a7d2a5bcc365f0ce3d/src/evaluation.py#L388C13-L392C84): the entire block of code (those 4 lines) should be inside the for loop.

Thanks for pointing that out!
